Introduce API Versions and Documentation

nils-wisiol opened the pull request:
This PR introduces API Version 2, that is a currently an alias for Version 1 and shall later be a stable copy of Version 1. Along the way, this PR cleans up the API definitions and re-organizes them. It also includes end-user documentation for API versioning.

Closes #137

peterthomassen commented:
I don't quite understand the approach. Based on the lifecycle section that you added in the docs, it seems like `v1` is the dev version and `v2` is production, but unreleased. How can that be?

Wouldn't we want to currently have `v1` with "release candiate"/"staging" status, and once we declare it stable have a follow up version `v2`?

Also, would it be worth considering `v0` for whatever we have before full launch, and at that point freeze the current state, naming in `v1`?

Also, we'd need to align the tests with whatever scheme we decide to use.

nils-wisiol commented:
Let's have an offline discussion about this, then report back with the results

nils-wisiol commented:
I dismissed the idea of a `v0` version and updated the docs to have a life cycle for each API version, starting from v1.

Also, I prepared code and docs for future versions.

nils-wisiol commented:
Before this is ready to review, we should decide how to handle the docs. readthedocs.io currently maintains a version of the docs for each branch, and I believe for each tag (we currently have none) of the desec-stack repository. We could therefore rely on readthedocs.io to maintain a API version 1, API version 2 documentation. The other option is to include all documentation in the master branch. Which way should we go? @peterthomassen

peterthomassen commented:
I prefer that each branch / tag carries the docs that are applicable to it.

nils-wisiol commented:
I updated the PR to reflect the tag-based approach. It has the downside that the life cycle document now cannot be placed inside the docs, or needs to be updated in old versions. I moved the status table in to README, keeping the remaining information in the docs. The tagging process is somewhat tedious, as tags need to be created and pushed separately, like so:

    git commit
    git tag -f apiv1
    git push
    git push -f origin apiv1

I'm afraid this can easily be forgotten, leading to outdated docs. An alternative way could be to use branches, but I'm not sure if that's any better.

On the other hand, the docs (in particular, the table of contents on [readthedocs.io](https://desec.readthedocs.io/en/latest/)) will not be cluttered with outdated information.

nils-wisiol commented:
README extension: https://github.com/desec-io/desec-stack/tree/20190222_api_versions#api-versions-and-road-map
readthedocs: https://desec.readthedocs.io/en/apiv1/#api-versions-and-life-cycle (notice the "apiv1")

nils-wisiol commented:
after much fiddling with the markdown table, this is finally ready for review :)

nils-wisiol commented:
This PR still has the problem that the unit tests only run for one set of API versions.

nils-wisiol commented:
Ready for further review.

- I introduced rest framework versioning together with django instance namespacing for URLs, [in a bit of an abuse](https://docs.djangoproject.com/en/2.2/topics/http/urls/#introduction) (see api/urls.py, where version 1 and version 2 are included as different instances of the same app).
- I added extensive documentation for how to use rest framework's `reverse()` in this setting in `/api/api/urls.py`.
- Default behavior for all linking is now to maintain the version that the user used to contact the API. If that is not provided or not available (let's say in the context of a django command), the code needs to define the behavior, otherwise an exception is raised.
- The unit tests are only run for version 1 of the API, this makes sense, as a new version also requires new tests. The fact that version 2 is a copy of version 1 as of now is immaterial.
- As a bit of (unrelated) sugar, I added a coverage report to the unit tests run on travis. We can later upgrade this to be shown on the readme or required to increase (not decrease) with every PR or what not.
- README now shows API version status, readthedocs.io shows API docs for the current and tagged versions. Updating a version now requires that the tag is newly set by using `git tag -f v1` and `git push -f origin v1`, **otherwise the docs will be out of date**.
- What should be up for debate is the fact that I decided [against all advise](https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning) that we keep the API version number in the URLs. I believe using versioning based on the accept header may be confusing for newcomers.

peterthomassen commented on api/api/urls.py:
```diff
@@ -1,38 +1,57 @@
+#
+#         r = RequestFactory().request(HTTP_HOST=settings.ALLOWED_HOSTS[0])
+#         r.version = 'v1'
+#         r.versioning_scheme = NamespaceVersioning()
```
How about writing a class for this, like `desecapi.AbstractRequest`? This would remove code duplication wherever this pattern is needed, and also the need to know about the details, like which versioning scheme is specified in the settings.

peterthomassen replied:
I suppose one could also retrieve `NamespaceVersioning` from `settings.py` and assign the property here dynamically.

nils-wisiol replied:
It's an option, but it's not needed right now. We do not use `resolve()` in django commands right now. Once we do, we can see if there is actual code duplication. Also, rest_framework is working on a better solution that does not rely on the `django.test.RequestFactory`.

peterthomassen commented on api/api/urls.py:
```diff
@@ -1,38 +1,57 @@
+#
 
+# IMPORTANT: specify default version last
+# if no other information is available, the last version will be used as default for reversing URLs
```
The "last" version, meaning the latest (highest number), or the last element of the list? I'm like 70:30 certain that you mean the latter, but maybe we can make this more clear by referring to the last list element explicitly.

nils-wisiol replied:
If in line "last" is replaced with "latest", the sentence does not make sense. Nevertheless, I replace it with the next push

peterthomassen commented on api/desecapi/tests/testregistration.py:
```diff
@@ -105,20 +108,23 @@ def test_multiple_registration_not_locked_same_ip_long_time(self):
-        send_account_lock_email(None, user)
+        r = RequestFactory().request(HTTP_HOST=settings.ALLOWED_HOSTS[0])
+        r.version = 'v1'
+        r.versioning_scheme = NamespaceVersioning()
```
See my comment in `urls.py`. Knowledge of the versioning scheme class is overspecific here, I think. A wrapper would solve this.

nils-wisiol replied:
It is very specific, and it needs to be, as the framework has no other source of knowing which versioning scheme is used. Since we're testing if there is an email sent, it's also not important which scheme we're using.

It's the only place where we need to provide a mocked request, which is why I do not want to wrap it.

peterthomassen replied:
The code _can_ figure out which versioning scheme is used, by looking at the settings. Anyways, I agree with your conclusion.

peterthomassen commented on api/desecapi/views.py:
```diff
@@ -297,14 +297,15 @@ class Root(APIView):
+                'logout': reverse('token-destroy', request=request),  # TODO change interface to token-destroy, too?
             })
         else:
+            print(request.version)
```
remove

nils-wisiol replied:
woops.

peterthomassen commented on api/desecapi/tests/testdomains.py:
```diff
@@ -11,22 +11,22 @@
 class UnauthenticatedDomainTests(APITestCase):
     def testExpectUnauthorizedOnGet(self):
-        url = reverse('domain-list')
+        url = reverse('v1:domain-list')
```
The version string is duplicated here many many times. How about importing a wrapped version of this, e.g. `from desec.test.utils import reverse`? We could have this function use some test-global state (as we have for the test user settings) to store the version number, and then just call the native `reverse()` with the first argument augmented.

nils-wisiol replied:
I considered and dismissed that idea. Each test is written for a specific version of the API, reusing tests is complicated and may be of limited use. Because we cannot decide today how to reuse tests, as we do not know how the future version may look like, it is fair to update all tests to their specific version. Admittedly, it looks hideous in the commit diff, but it's not a lot of work with search and replace if it needs to be changed later. Again, today, I do not see a scenario where it is useful to change all tests to another API version.

peterthomassen commented on api/desecapi/tests/testdomains.py:
```diff
@@ -1,7 +1,7 @@
 from django.urls import reverse
```
In `urls.py`, you wrote: 

> Always use `rest_framework.reverse.reverse`, do not directly use `django.urls.reverse`

Should we follow this recommendation here (and in the rest of the code)?

nils-wisiol replied:
yup. Fixed with next push

peterthomassen commented on docs/lifecycle.rst:
```diff
@@ -0,0 +1,31 @@
+API Versions and Life Cycle
```
This section is generally clear, but I think we can smoothen the language a bit. If you don't mind, I'd proofread this and add a commit with suggestions, which you can either accept or pick from. Ok for you?

nils-wisiol replied:
Go nuts, commit your changes, and let me know which id to cherry-pick

peterthomassen replied:
See https://github.com/desec-io/desec-stack/pull/140/commits/d5b82b5e986382548ee87acf7ab634aa4c0c14be

peterthomassen commented on api/desecapi/urls/version_1.py:
```diff
@@ -0,0 +1,61 @@
+    url(r'^donation/', views.DonationList.as_view(), name='donation'),
+
+    # Unlock endpoints
+    url(r'^unlock/user/(?P<email>.+)$', views.unlock, name='unlock/byEmail'),
```
I just realized that this causes PII to be logged (in this case, email address). This is considered bad practice. Also, there may be issues with URI encoding edge cases etc.

I created an issue for this (#161), but I don't think it needs to be in scope for launch.

peterthomassen commented on api/desecapi/urls/version_1.py:
```diff
@@ -0,0 +1,61 @@
+    # Token management
+    url(r'^token/login/$', views.TokenCreateView.as_view(), name='login'),
+    url(r'^token/logout/$', views.TokenDestroyView.as_view(), name='logout'),
+    url(r'^', include('djoser.urls.authtoken')),  # note: this is partially overwritten by the two lines above
```
This line has now moved up a bit, in comparison to the previous arrangement (https://github.com/desec-io/desec-stack/blob/224eeb9fef477332dc5eb3e67702d5fc6707cb37/api/api/urls.py#L18). Does this have any implications, and if so, what's the reasoning behind the change?

nils-wisiol replied:
It's rearranged and documented with line comments

peterthomassen replied:
Yes, of course. Does it have any implications though (for example, has the way that routes can shadow each other changed)?

nils-wisiol replied:
imho it's functionally equivalent

peterthomassen commented on api/requirements.txt:
```diff
@@ -8,3 +8,4 @@ requests~=2.21.0
 uwsgi~=2.0.0
 django-nocaptcha-recaptcha==0.0.20  # updated manually
 djangorestframework-bulk~=0.2.0
+coverage~=4.5.3
```
Is there a way to only have such dependencies when testing, but not in the final build?

This does not need to be solved now, but if we think about hardening later, we should make sure that stuff not used at runtime does not get built into the image.

nils-wisiol replied:
It's possible, but it introduces additional complexity (such as waiting for coverage to be installed with every test run or having a second requirements.txt file). I believe as of now, the disadvantages outweigh the advantages. (Also considering that we do not even install the latest version of djangorestframework and use the abandoned django-nocaptcha-recaptcha and djangorestframework-bulk)

nils-wisiol commented:
I noticed while cleaning up the import statements that we're using the outdated `url()` function to specify the URL configuration. `url()` is an alias for `re_path()`, I updated the code accordingly with a46eb30ffd29ef02a6314f9d96b7e9add4d41119. Also, I removed regex where it was not necessary.

peterthomassen commented:
I added my docs modifications. Once you reviewed them, I think we're good to go (modulo my question on the order in `urls.py`).

I suggest to only keep the `import` cleanup commit and the coverage integration commit separate, and squash the rest.

nils-wisiol commented:
imho ready to merge
